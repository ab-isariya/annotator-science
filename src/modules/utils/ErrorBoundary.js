import React from 'react';
import PropTypes from 'prop-types';

import Alert from '@modules/ui/Alert';

/**
 * As of v16.2.0, there's no way to turn a functional component
 * into an error boundary.
 */
class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = {error: null, errorInfo: null};
  }

  componentDidCatch(error, errorInfo) {
    this.setState({
      error: error,
      errorInfo: errorInfo
    });
  }

  render() {
    if (this.state.errorInfo) {
      return (
        <div className="container mx-auto my-8 font-inter">
          <h1 className="text-2xl">Something went wrong</h1>

          <Alert level="error">
            {this.state.error && this.state.error.toString()}
          </Alert>

          <details className="py-6 px-0 whitespace-pre-wrap	">
            <p>{this.state.errorInfo.componentStack}</p>
          </details>
        </div>
      );
    }
    // Render children if there's no error
    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  //React node children
  children: PropTypes.node
};

export default ErrorBoundary;
