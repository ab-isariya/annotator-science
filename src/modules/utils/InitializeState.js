/* istanbul ignore file */
import {setInitialState as SpotlightInitialize} from '@document/data/useSpotlight';
import {setInitialState as LeftSidebarInitialize} from '@document/LeftSidebar';
import {setInitialState as DocumentViewInitialize} from '@document/data/useDocument';
import {setInitialState as UseAnnotationsInitialize} from '@document/data/useAnnotations';

export default () => {
  SpotlightInitialize();
  LeftSidebarInitialize();
  DocumentViewInitialize();
  UseAnnotationsInitialize();
};
