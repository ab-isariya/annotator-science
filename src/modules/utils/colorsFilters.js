/* istanbul ignore file */
/**
 * @file These are the needed steps to convert colors in our SVG icons
 *
 * to be used as <Delete style={{filter: `${toBlack} ${blackToWhite}`}} />
 * where <Delete /> is an svg of any color
 *
 * Steps created with this generator: https://codepen.io/sosuke/pen/Pjoqqp
 */

// Any color to black
export const toBlack = 'brightness(0) saturate(100%)';

// From black to white
export const blackToWhite =
  'invert(100%) sepia(91%) saturate(38%) hue-rotate(321deg) brightness(110%) contrast(110%)';

// Black to grey500
export const blackToGrey500 =
  'invert(42%) sepia(23%) saturate(354%) hue-rotate(182deg) brightness(94%) contrast(85%)';

export const blackToGrey200 =
  'invert(95%) sepia(2%) saturate(3211%) hue-rotate(182deg) brightness(91%) contrast(79%)';

// Black to red500
export const blackToRed500 =
  'invert(18%) sepia(26%) saturate(6877%) hue-rotate(319deg) brightness(91%) contrast(96%)';

export const blackToBlue500 =
  ' invert(39%) sepia(17%) saturate(3351%) hue-rotate(200deg) brightness(95%) contrast(89%)';
