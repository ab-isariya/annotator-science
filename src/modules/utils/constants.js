import {colors} from '@ui/theme';
import {ReactComponent as IconDocx} from '@assets/svgs/IconDocx.svg';
import {ReactComponent as IconPdf} from '@assets/svgs/IconPdf.svg';

/*
 * Entity/Annotation Status enum
 * See tat-api/annotate_api/chalicelib/database/utils.py
 */
export const AnnotationStatus = {
  ACCEPTED: 'ACCEPTED',
  NOT_REVIEWED: 'NOT_REVIEWED', //DEFAULT
  REJECTED: 'REJECTED',
  MANUAL: 'MANUAL'
};

/** Labels for Entity/Annotation status' */
export const AnnotationStatusLabel = {
  ACCEPTED: 'Accepted',
  REJECTED: 'Rejected',
  NOT_REVIEWED: 'To Review',
  MANUAL: 'Added'
};

//NOTE(Rejon): Progress Type aggregations aren't sorted on the backend,
//             this array also helps controls the expected order for Progress Types
export const sortedAnnotationStatus = [
  AnnotationStatus.ACCEPTED,
  AnnotationStatus.REJECTED,
  AnnotationStatus.NOT_REVIEWED,
  AnnotationStatus.MANUAL
];

/** Color codes specifically used for PieChart components. */
export const annotationStatusColors = {
  ACCEPTED: colors.green['400'],
  REJECTED: colors.red['600'],
  NOT_REVIEWED: colors.grey['200'],
  MANUAL: colors.blue['500'],
  ZERO: colors.grey['50'] // no annotations
};

// Color codes for statuses rendered using only Text
export const altAnnotationStatusColors = {
  ACCEPTED: 'text-green-500',
  REJECTED: 'text-red-600',
  NOT_REVIEWED: 'text-grey-500',
  MANUAL: 'text-black'
};

//Annotation type tags based on Kepler spec.
export const annotationTypes = {
  ANATOMY: 'anatomy & physiology',
  CELLBIOLOGY: 'cell biology',
  CHEMICALS: 'chemicals & drugs',
  DEVICES: 'medical devices',
  CONDITIONS: 'medical conditions',
  GENETICS: 'genetics',
  PROCEDURES: 'medical procedures',
  VIRUSES: 'species & viruses',
  CONTEXT: 'context',
  UNASSIGNED: 'unassigned'
};

/*
 * Document processing state
 * See tat-api/annotate_api/chalicelib/database/utils.py
 *
 * Sorted values used in src/modules/document/table/sortFunctions.js
 */
export const ProcessingState = {
  CREATED: 'CREATED',
  UPLOADED: 'UPLOADED',
  QUEUED: 'QUEUED',
  FAILED: 'FAILED',
  PROCESSING: 'PROCESSING',
  PROCESSED: 'PROCESSED'
};

//Processing State sorted
export const sortedProcessingStatus = [
  'CREATED',
  'UPLOADED',
  'QUEUED',
  'PROCESSING',
  'PROCESSED'
];

/*
 * Annotation confidence scores based on Kepler scoring rubric
 */
export const ConfidenceScore = {
  very_high: {
    label: 'Very High',
    max: 1,
    min: 0.99,
    tooltipText: 'Very High Confidence',
    score_cat: 'VERY_HIGH'
  },
  high: {
    label: 'High',
    max: 0.98,
    min: 0.9,
    tooltipText: 'High Confidence',
    score_cat: 'HIGH'
  },
  moderate: {
    label: 'Moderate',
    max: 0.89,
    min: 0.7,
    tooltipText: 'Moderate Confidence',
    score_cat: 'MODERATE'
  },
  low: {
    label: 'Low',
    max: 0.69,
    min: 0.5,
    tooltipText: 'Low Confidence',
    score_cat: 'LOW'
  },
  very_low: {
    label: 'Very Low',
    max: 0.49,
    min: 0,
    tooltipText: 'Very Low Confidence',
    score_cat: 'VERY_LOW'
  }
};

//NOTE(Rejon): Confidence Type aggregations don't sort on the backend,
//             this controls the expected sort order if confidence type
//             is ever listed together.
export const ConfidenceTypeOrder = [
  'very_high',
  'high',
  'moderate',
  'low',
  'very_low'
];

// Keys that correspond to the backend model for filtering Annotations
// based on the client side modules they're used in.
export const AnnotationFilterKeys = {
  ModelConfidence: 'entity_p',
  Progress: 'status',
  Types: 'tag',
  Concepts: 'canonical_id'
};

//Possible view states for the document view.
export const DocumentViewTypes = {
  TABLE: 'table',
  DOCUMENT: 'document'
};

//Possible Document Export table options
export const DocumentExportTypes = {
  TSV: 'TSV',
  CSV: 'CSV',
  XLSX: 'XLSX'
};

//Lenght that a snippet should be.
export const SNIPPET_CLAMP = 25;

//Right sidebar view modes.
export const RightSidebarModes = {
  REVIEW: 'REVIEW',
  ADD: 'ADD',
  EDIT: 'EDIT'
};

//Supported upload file type icons.
export const FileTypes = {
  'docx': IconDocx,
  'pdf': IconPdf
};
