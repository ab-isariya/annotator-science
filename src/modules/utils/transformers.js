import DateTime from 'luxon/src/datetime';

/**
 * Returns our UTC datetimes in a given format
 *
 * @param {String} datetimeStr
 * @param {Object} format - valid object to Luxon
 * https://moment.github.io/luxon/docs/class/src/datetime.js~DateTime.html
 * @return {string}
 */
export function datetimeToHuman(datetimeStr, format) {
  return DateTime.fromISO(datetimeStr, {zone: 'utc'})
    .toLocal()
    .toLocaleString(format);
}

/**
 * Returns the current datetime as an ISO.
 *
 * NOTE! It removes the Z from the end because AWS RDS doesn't like keeping the Z
 * in it's records.
 *
 * @return {string}
 */
export function nowISOTime() {
  return new Date().toISOString().slice().slice(0, -1); //NOTE(Rejon): Drop the Z!;
}

/**
 * Returns size in human readable units
 *
 * @param {Number} sizeInBytes - size in bytes
 * @return {String} human readable size and the unit
 */
export function humanFileSize(sizeInBytes) {
  const i =
    sizeInBytes === 0 ? 0 : Math.floor(Math.log(sizeInBytes) / Math.log(1024));
  return (
    (sizeInBytes / Math.pow(1024, i)).toFixed(1) +
    ['B', 'kB', 'MB', 'GB', 'TB'][i]
  );
}

/**
 * Rounding function that avoids the problem of decimal of 5 rounded down
 * instead of up
 *
 * @param {Number} value
 * @param {Number} decimals
 * @return {number}
 */
export function round(value, decimals) {
  return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals);
}

/**
 * Formats a number (integer or float) into a string with comma
 * separators. Also handles decimals for floats.

 * @param {number} amount - Number to be formatted
 * @return {string}
 */
export const CommaFormatted = (amount) => {
  var delimiter = ','; // replace comma if desired
  var a = String(amount).split('.', 2);
  var d = a[1];
  var i = parseInt(a[0]);
  if (isNaN(i)) {
    return '';
  }
  var minus = '';
  if (i < 0) {
    minus = '-';
  }
  i = Math.abs(i);
  var n = String(i);
  a = [];
  while (n.length > 3) {
    var nn = n.substr(n.length - 3);
    a.unshift(nn);
    n = n.substr(0, n.length - 3);
  }
  if (n.length > 0) {
    a.unshift(n);
  }
  n = a.join(delimiter);
  if (d === undefined || d.length < 1) {
    amount = n;
  } else {
    amount = n + '.' + d;
  }
  amount = minus + amount;
  return amount;
};

/**
 * Formats a number to a shortend place (ie. 1,000 becomes 1.0k)
 * //From: https://stackoverflow.com/questions/9461621/format-a-number-as-2-5k-if-a-thousand-or-more-otherwise-900
 * @constructor
 * @param {number} num - Number to be formatted.
 * @param {number} digits - Amount of digits to support when formatting (ie. 10,080 with 3 digits -> 10.1k with 4 digits -> 10.05k)
 * @return {string}
 */
export const nFormatter = (num, digits) => {
  var si = [
    {value: 1, symbol: ''},
    {value: 1e3, symbol: 'k'},
    {value: 1e6, symbol: 'M'},
    {value: 1e9, symbol: 'G'},
    {value: 1e12, symbol: 'T'},
    {value: 1e15, symbol: 'P'},
    {value: 1e18, symbol: 'E'}
  ];
  var rx = /\.0+$|(\.[0-9]*[1-9])0+$/;
  var i;
  for (i = si.length - 1; i > 0; i--) {
    if (num >= si[i].value) {
      break;
    }
  }
  return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
};

/**
 * Formatter that combines nFormatter and CommaFormatted
 * based on a limit. If num is less than the limit, it'll be comma formatted,
 * else, it'll be nFormatted.
 * NOTE: limit is 1000 by default
 *
 * @param {number} num - Number to be formatted.
 * @param {number} digits - Number of digits to show for nFormatted
 * @param {number} limit (Default: 1000) - Limit to change the format from CommaSeperated into nFormatted.
 * @return {string}
 */
export const NumberCommaSuffix = (num, digits, limit = 1000) => {
  if (isNaN(num)) {
    return '';
  }

  if (num <= limit) {
    return CommaFormatted(String(num));
  } else {
    return nFormatter(num, digits);
  }
};
