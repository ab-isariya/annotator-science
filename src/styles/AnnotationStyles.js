// See https://www.figma.com/file/DdbiQRZHR33IX83pYJwsNU/2021-Master-Components?node-id=3106%3A23
// for specs on tag styling

//Annotation Background Styles
//NOTE: Text and BG are seperated due to rejected state having "text-black" class override
export const annBGStyles = {
  // Anatomy & Physiology
  'anatomy & physiology': 'hover:bg-red-50',
  // Cell biology
  'cell biology': 'hover:bg-green-50',
  // Chemicals & Drugs
  'chemicals & drugs': 'hover:bg-orange-50',
  // Medical devices
  'medical devices': 'hover:bg-purple-50',
  // Medical Conditions
  'medical conditions': 'hover:bg-blue-50',
  // Genetics and Mutation
  genetics: 'hover:bg-green-50',
  // Medical Procedures
  'medical procedures': 'hover:bg-violet-50',
  // Species & Viruses
  'species & viruses': 'hover:bg-cyan-50',
  // Context
  context: 'hover:bg-grey-50',
  unassigned: 'hover:bg-grey-50'
};

//Annotation Text Styles
//NOTE: Text and BG are seperated due to rejected state having "text-black" class override
export const annTextStyles = {
  // Anatomy & Physiology
  'anatomy & physiology': 'text-red-500',
  // Cell biology
  'cell biology': 'text-green-cellBiology',
  // Chemicals & Drugs
  'chemicals & drugs': 'text-orange-500',
  // Medical devices
  'medical devices': 'text-purple-500',
  // Medical Conditions
  'medical conditions': 'text-blue-500',
  // Genetics and Mutation
  genetics: 'text-green-500 hover:text-green-600',
  // Medical Procedures
  'medical procedures': 'text-violet-500',
  // Species & Viruses
  'species & viruses': 'text-cyan-500',
  // Context
  context: 'text-grey-500'
};

//Annotation text-decoration styles for things like strikethroughs
export const annDecorationStyles = {
  // Anatomy & Physiology
  'anatomy & physiology': 'decoration-color-red-500',
  // Cell biology
  'cell biology': 'decoration-color-green-cellBiology',
  // Chemicals & Drugs
  'chemicals & drugs': 'decoration-color-orange-500',
  // Medical devices
  'medical devices': 'decoration-color-purple-500',
  // Medical Conditions
  'medical conditions': 'decoration-color-blue-500',
  // Genetics and Mutation
  genetics: 'decoration-color-green-600',
  // Medical Procedures
  'medical procedures': 'decoration-color-violet-500',
  // Species & Viruses
  'species & viruses': 'decoration-color-cyan-500',
  // Context
  context: 'decoration-color-grey-500'
};

//Activate Annotation Styles
export const annActiveStyles = {
  // Anatomy & Physiology
  'anatomy & physiology': 'border-red-500 bg-red-50',
  // Cell biology
  'cell biology': 'border-green-cellBiology bg-green-50',
  // Chemicals & Drugs
  'chemicals & drugs': 'border-orange-500 bg-orange-50',
  // Medical devices
  'medical devices': 'border-purple-500 bg-purple-50',
  // Medical Conditions
  'medical conditions': 'border-blue-500 bg-blue-50',
  // Genetics and Mutation
  genetics: 'border-green-600 bg-green-50',
  // Medical Procedures
  'medical procedures': 'border-violet-500 bg-violet-50',
  // Species & Viruses
  'species & viruses': 'border-cyan-500 bg-cyan-50',
  // Context
  context: 'border-grey-500 bg-grey-50'
};

export const statusStyles = {
  ACCEPTED: {
    bg: 'bg-green-400',
    text: 'text-green-400'
  },
  REJECTED: {
    bg: 'bg-red-600',
    text: 'text-red-600'
  },
  NOT_REVIEWED: {
    bg: 'bg-transparent',
    text: 'text-white'
  }
};
