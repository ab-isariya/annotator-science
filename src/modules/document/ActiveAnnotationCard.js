import PropTypes from 'prop-types';

import classNames from 'classnames';
import findIndex from 'lodash/findIndex';
import styled from 'styled-components';
import {DateTime} from 'luxon';
import produce from 'immer';

import Button from '@ui/Button';
import Card from '@ui/Card';
import ConfidenceScoreLabel from '@ui/ConfidenceScoreLabel';
import ReviewToggle from '@document/ReviewToggle';
import Tooltip from '@ui/Tooltip';
import {AnnotationStatus} from '@utils/constants';
import {rightSidebarState, useAnnotations, useDocument} from '@document';
import {colors} from '@ui/theme';
import {nowISOTime, datetimeToHuman} from '@utils/transformers';
import {updateAnnotation} from '@document/api';
import {useUser} from '@user';

import typeStyles from '@styles/TypeStyles';
import {statusStyles} from '@styles/AnnotationStyles';
import {ReactComponent as HorizontalSort} from '@assets/svgs/Horizontal_Sort.svg';
import {ReactComponent as TrashCan} from '@assets/svgs/TrashCan.svg';

//Left border span color based on annotation status.
//Pulls colors from theme.js file
export const StatusBorder = styled.span`
 background: ${({status}) =>
    status === AnnotationStatus.NOT_REVIEWED
      ? 'transparent'
      : status === AnnotationStatus.ACCEPTED
        ? colors.green['400']
        : status === AnnotationStatus.REJECTED
          ? colors.red['600']
          : 'transparent'}
     
  transform: ${({status}) =>
    status !== AnnotationStatus.NOT_REVIEWED
      ? 'translateX(-8px)'
      : 'translateX(0px)'};

transition: all 0.2s ease-in-out;
`;

const ActiveAnnotationCard = (props) => {
  const {
    // text,
    entity_p,
    tag,
    id,
    status,
    canonical_name,
    canonical_id,
    score_cat,
    datetime_modified,
    // datetime_created,
    className
  } = props;

  const {data: user} = useUser();
  const {activeAnnotation} = rightSidebarState.useValue();
  const {data: annotations, mutate: mutateAnnotations} = useAnnotations();
  const {data: document, mutate: mutateDocument} = useDocument();

  if (status === AnnotationStatus.MANUAL) {
    return (
      <Card className={`mb-2.5 relative overflow-hidden mt-7 ${className}`}>
        <span
          className="h-full left-0 top-0 absolute bg-holofoil_blue"
          style={{width: '6px'}}></span>
        <div
          className={`font-sans font-bold mb-2 text-${typeStyles[tag]['textColor']}`}>
          {canonical_name}
        </div>
        <div className="flex justify-between flex-col">
          <div className="flex flex-row mb-2">
            <Tooltip
              content={typeStyles[tag]['label']}
              arrow={false}
              followCursor={true}
              size="small">
              {typeStyles[tag]['icon']}
            </Tooltip>
            <Tooltip
              content={canonical_name}
              arrow={false}
              followCursor={true}
              size="small">
              <span className="block ml-2 text-grey-500 underline font-normal font-mono text-sm">
                {canonical_id}
              </span>
            </Tooltip>
          </div>
          <p className="text-xs text-grey-300 mb-2">
            Last Edited on{' '}
            {datetimeToHuman(datetime_modified, DateTime.DATE_FULL)}
          </p>
          <div className="flex flex-row items-center justify-start">
            <Button tertiaryIcon className="w-8 h-8 mr-2">
              <Tooltip
                content={'Change Linked Concept'}
                size={'small'}
                placement="bottom">
                <HorizontalSort className="text-grey-500" />
              </Tooltip>
            </Button>
            <Button tertiaryIcon className="w-8 h-8">
              <Tooltip
                content={'Delete Annotation'}
                size={'small'}
                placement="bottom">
                <TrashCan className="text-grey-500" />
              </Tooltip>
            </Button>
          </div>
        </div>
      </Card>
    );
  }

  const reviewAnnotation = async (newStatus) => {
    //Check if new status is the same as our current status.
    //If it is set the new status to NOT_REVIEW
    const _newStatus =
      newStatus === status
        ? AnnotationStatus.NOT_REVIEWED
        : AnnotationStatus[newStatus];

    //Find the index of this annotation in the documents annotation.
    const annotationIndex = findIndex(annotations, (o) => o.id === id);

    //NOTE(Rejon): Remove the annotation we're updating from preflight,
    //             so we can attach a preFlight annotation during mutation.
    let preFlightAnnotations = [...annotations];

    //Set up our preFlight annotation
    const preFlightAnnotation = {
      ...props,
      node: null,
      datetime_reviewed: nowISOTime(),
      status: _newStatus
    };

    //Set up our preFlight aggregations
    let preflightProgressCount = document.aggregations.progress.count; //Store the count, we're going to edit it

    const indexOfOldStatus = findIndex(
      //Find the index in our aggregations of the old status
      document.aggregations.progress.status,
      (o) => o === status
    );

    const indexOfNewStatus = findIndex(
      //Find the index in our aggregations of the new status
      document.aggregations.progress.status,
      (o) => o === _newStatus
    );

    //Subtract 1 from the old status, Add 1 to the new status
    preflightProgressCount[indexOfOldStatus] =
      preflightProgressCount[indexOfOldStatus] - 1;

    preflightProgressCount[indexOfNewStatus] =
      preflightProgressCount[indexOfNewStatus] + 1;

    //Set up preFlight aggregation object
    const preFlightAggregation = {
      ...document.aggregations,
      progress: {
        count: preflightProgressCount,
        ...document.aggregations.progress
      }
    };

    //Update our preFlightAnnotations annotation at the specific index.
    //NOTE(Rejon): I put preFlight: true here instead of the object above,
    //             because we want to avoid updating activeAnnotationState twice
    //             when preFlight is automatically removed by swr refresh.
    preFlightAnnotations[annotationIndex] = {
      ...preFlightAnnotation,
      preFlight: true
    };

    //If this annotation is the active annotation, update with the new data.
    if (activeAnnotation.id === id) {
      rightSidebarState.set((prev) =>
        produce(prev, (updated) => {
          updated.activeAnnotation = preFlightAnnotation;
        })
      );
    }

    //Optimistically update the UI immediately
    mutateDocument(
      {
        ...document,
        aggregations: preFlightAggregation,
        annotations: preFlightAnnotations
      },
      false
    );

    //Mutate annotations in-case of filtering immediately.
    //NOTE(Rejon): When filtering a seperate SWR hook is used for the query.
    //             So this is necessary to avoid the document not updating annotations when filtering.
    mutateAnnotations(preFlightAnnotations, false);

    //Update our annotations on the backend.
    const {annotations: newAnnotations, aggregations} = await updateAnnotation(
      user.id,
      user.accessToken,
      document.id,
      [preFlightAnnotation]
    );

    //Remove preflight from the annotation in the annotations array.
    preFlightAnnotations[annotationIndex] = newAnnotations[0];

    //Update our UI with new real data, revalidate endpoint.
    mutateDocument({
      ...document,
      annotations: preFlightAnnotations,
      aggregations
    });

    //Mutate annotations in-case of filtering with true data.
    //NOTE(Rejon): When filtering a seperate SWR hook is used for the query.
    //             So this is necessary to avoid the document not updating annotations when filtering.
    mutateAnnotations(preFlightAnnotations);
  };

  //use classnames for animation pulse
  const pulseClass = classNames({
    'animate-redpulse': status === AnnotationStatus.REJECTED,
    'animate-greenpulse': status === AnnotationStatus.ACCEPTED
  });

  return (
    <Card
      className={`mb-2.5 relative overflow-hidden mt-7 ${className} ${pulseClass}`}>
      <StatusBorder
        status={status}
        className={`status-border absolute h-full left-0 top-0 ${statusStyles[status].bg}`}
        style={{width: '6px'}}></StatusBorder>
      <ReviewToggle
        className="mb-3"
        status={status}
        reviewAnnotation={reviewAnnotation}
      />
      <div
        className={`font-sans font-bold mb-2 text-${typeStyles[tag]['textColor']}`}>
        {canonical_name}
      </div>
      <div className="flex justify-between">
        <div className="flex flex-row">
          <Tooltip
            content={typeStyles[tag]['label']}
            arrow={false}
            followCursor={true}
            size="small">
            {typeStyles[tag]['icon']}
          </Tooltip>
          <Tooltip
            content={canonical_name}
            arrow={false}
            followCursor={true}
            size="small">
            <span className="block ml-2 text-grey-500 underline font-normal font-mono text-sm">
              {canonical_id}
            </span>
          </Tooltip>
        </div>
        <ConfidenceScoreLabel type={score_cat} score={entity_p} />
      </div>
    </Card>
  );
};

ActiveAnnotationCard.propTypes = {
  // Annotation text displayed
  text: PropTypes.string.isRequired,
  //Unique ID for the annotation
  id: PropTypes.string,
  //Kepler model confidence score
  entity_p: PropTypes.number,
  //Tag type of the annotation
  tag: PropTypes.string.isRequired,
  //Review status of the annotation (See AnnotationStatus in constants.js)
  status: PropTypes.string.isRequired,
  //Canonical name of the annotation
  canonical_name: PropTypes.string,
  //Canonical ID of the annotation
  canonical_id: PropTypes.string,
  //Last datetime of the annotation being modified
  datetime_modified: PropTypes.string,
  //datetime of the annotation's creation
  datetime_created: PropTypes.string,
  //Constant all caps string of the confidence score category
  score_cat: PropTypes.string,
  // Tailwind Classes
  className: PropTypes.string
};

export default ActiveAnnotationCard;
