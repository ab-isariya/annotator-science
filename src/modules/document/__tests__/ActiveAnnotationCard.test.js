import {cache} from 'swr';

beforeEach(() => cache.clear());

import ActiveAnnotationCard, {StatusBorder} from '../ActiveAnnotationCard';
import {statusStyles} from '@styles/AnnotationStyles';
import {AnnotationStatus} from '../../utils/constants';

jest.mock('@okta/okta-react', () => ({
  useOktaAuth: () => {
    return {
      authState: {isAuthenticated: true},
      authService: {handleAuthentication: jest.fn()},
      oktaAuth: {
        tokenManager: {
          on: jest.fn()
        }
      }
    };
  }
}));

jest.mock('react-router-dom', () => ({
  useParams: () => ({
    docID: 1
  })
}));

describe('<ActiveAnnotationCard/>', () => {
  it('renders without crashing', () => {
    const mockAnnotation = {
      tag: 'chemicals & drugs',
      score_cat: 'VERY_HIGH',
      canonical_id: 'UMLS:C0042212',
      canonical_name: 'Vaccines, Inactivated',
      text: 'inactivated vaccine',
      status: 'NOT_REVIEWED',
      datetime_modified: '2021-06-24T16:31:03.219513',
      id: '1',
      entity_p: 0.9997294545173645,
      datetime_created: '2021-06-24T16:31:03.219513',
      end: '77',
      start: '58',
      datetime_reviewed: null,
      node: null
    };

    shallow(<ActiveAnnotationCard {...mockAnnotation} />);
  });

  it('Renders as a Manual Annotation Card', () => {
    const mockAnnotation = {
      tag: 'chemicals & drugs',
      score_cat: 'VERY_HIGH',
      canonical_id: 'UMLS:C0042212',
      canonical_name: 'Vaccines, Inactivated',
      text: 'inactivated vaccine',
      status: 'MANUAL',
      datetime_modified: '2021-06-24T16:31:03.219513',
      id: '1',
      entity_p: 0.9997294545173645,
      datetime_created: '2021-06-24T16:31:03.219513',
      end: '77',
      start: '58',
      datetime_reviewed: null,
      node: null
    };

    const tree = shallow(<ActiveAnnotationCard {...mockAnnotation} />);

    expect(tree).toMatchSnapshot();
  });

  it('should render a Status Border', () => {
    const mockAnnotation = {
      tag: 'chemicals & drugs',
      score_cat: 'VERY_HIGH',
      canonical_id: 'UMLS:C0042212',
      canonical_name: 'Vaccines, Inactivated',
      text: 'inactivated vaccine',
      status: 'NOT_REVIEWED',
      datetime_modified: '2021-06-24T16:31:03.219513',
      id: '1',
      entity_p: 0.9997294545173645,
      datetime_created: '2021-06-24T16:31:03.219513',
      end: '77',
      start: '58',
      datetime_reviewed: null,
      node: null
    };

    const _StatusBorder = (
      <StatusBorder
        status={'NOT_REVIEWED'}
        className={`status-border absolute h-full left-0 top-0 ${statusStyles['NOT_REVIEWED'].bg}`}
        style={{width: '6px'}}></StatusBorder>
    );

    const wrapper = shallow(<ActiveAnnotationCard {...mockAnnotation} />);

    expect(wrapper.contains(_StatusBorder)).toBe(true);
  });
});

describe('<StatusBorder/>', () => {
  it('renders status border as Accepted', () => {
    const tree = shallow(<StatusBorder status={AnnotationStatus.ACCEPTED} />);

    expect(tree).toMatchSnapshot();
  });

  it('renders status border as Not Reviewed', () => {
    const tree = shallow(
      <StatusBorder status={AnnotationStatus.NOT_REVIEWED} />
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders status border as Rejected', () => {
    const tree = shallow(<StatusBorder status={AnnotationStatus.REJECTED} />);
    expect(tree).toMatchSnapshot();
  });
});
