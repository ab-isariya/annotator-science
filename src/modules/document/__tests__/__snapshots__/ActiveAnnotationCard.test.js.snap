// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActiveAnnotationCard/> Renders as a Manual Annotation Card 1`] = `ShallowWrapper {}`;

exports[`<StatusBorder/> renders status border as Accepted 1`] = `ShallowWrapper {}`;

exports[`<StatusBorder/> renders status border as Not Reviewed 1`] = `ShallowWrapper {}`;

exports[`<StatusBorder/> renders status border as Rejected 1`] = `ShallowWrapper {}`;
